kore/src/Kore/Step/Step.hs

```diff
@@ -303,34 +268,30 @@ applyUnifiedRule
 
     -> PredicateSubstitution Object variable
     -- ^ Initial conditions
-    -> UnifiedRule variable
-    -- ^ Non-normalized final configuration
-    -> BranchT unifier
-        (ExpandedPattern Object variable)
-applyUnifiedRule
+    -> PredicateSubstitution Object variable
+    -- ^ Unification conditions
+    -> BranchT unifier (OrOfPredicateSubstitution Object variable)
```

Comment:
I didn't think carefully about this, but why do you need BranchT here? Is it only to avoid a `lift` in the caller?

The same question for `finalizeAppliedRule`. Actually, since we're restarting to work with `OrOfStuff`, do we still need BranchT in this file?

Reply:
> Is it only to avoid a `lift` in the caller?

Yes. This keeps all the code in the caller at the same level of abstraction.

> Actually, since we're restarting to work with `OrOfStuff`, do we still need BranchT in this file?

Yes. Your comments on another recent pull request made me realize that we cannot always satisfy the internal invariants of `MultiOr`. For example, when applying rules, we need to branch on things that might not be `TopBottom`, and we may need to keep branches "alive" even when they return something that `isBottom`, such as when the right-hand side of a rule is bottom.

Reply:
> This keeps all the code in the caller at the same level of abstraction.

Replying to my own comment: This won't be necessary anymore when we clean up the issues around defining an `instance MonadUnify (BranchT unifier)`.